# Lecture 2 Notes

## The Science of Algorithms

Computer science can be seen as the study of **algorithms**.

An algorithm is a precise step-by-step description of how to solve a problem.

Before we dive into the details of programming, lets take a look at couple of
interesting algorithms.

## A Greatest Common Divisor Algorithm

Given two positive integers, like 51 and 90, what's the biggest integer that
divides both numbers? Since $51 = 3 \cdot 17$ and $90 = 2 \cdot 3 \cdot 3 \cdot
5$, the answer is 3. We call this the **greatest common divisor**, or **GCD**,
of 51 and 90.

Try these questions:

- What's the GCD of 15 and 39?
- What's the GCD of 7 and 5?

What method did you use to answer the questions? Would it work efficiently for
any pair of numbers?

### Euclid's Algorithm in Pseudocode

A famous algorithm for calculating GCDs is [Euclid's
algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm). It was first
described in 300 BC in [Euclid's
Elements](https://en.wikipedia.org/wiki/Euclid%27s_Elements), and so is over
2000 years old.

[Euclid's algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm) says
that to find the greatest common divisor of two positive integers, do the
following:

1. if the numbers are the same, stop: their GCD is the number itself
2. if they are different, replace the bigger number with the difference between
   it and the smaller number
3. go to step 1

When we write an algorithm in English we call it **pseudocode**. We will often
use pseudocode when talking about algorithms.

### Example
Calculate the GCD of 15 and 39 using [Euclid's
algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm). Show each step.

We start with 15 and 39. Step 1 of the algorithm says to check if the numbers
are the same. If they are the same, we're done. But in this case they're not, so
we go to line 2 which says to replace the bigger number with difference between
it and the other number. So 15 and 39 becomes 15 and 39-15=24, or just 15 and
24.

Now we repeat steps 1 and 2 until the numbers are the same:

```
39 15
24 15
9 15
9 6
3 6
3 3
```

When the numbers are the same [Euclid's
algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm) tells us to stop,
and so the GCD of 39 and 15 is 3.

Computer scientists have studied questions like these about [Euclid's
algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm):
- **Does it always return the correct answer?** Are there numbers that it
  doesn't work for, or, say, get's stuck in an infinite loop with? Can we
  *prove* it always works correctly?
    - If you're curious, you can prove [the correctness of Euclid's
      algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm#Proof_of_validity)
      using some number theory.
- **How efficient is this algorithm?** Can we make it faster? Or perhaps is
  there a different algorithm that calculates GCDs more quickly?
    - If you're curious, check out [the performance of Euclid's
      algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm#Algorithmic_efficiency).


### Euclid's Algorithm in Python

Here is a Python implementation of [Euclid's
algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm) ... click on the
arrow at the top left of the cell to run it:

In [4]:
# tell the user what the program does
print('Please enter two integers.')
print("I will calculate their GCD using Euclid's algorithm")
print()

# get the input numbers from the user (as strings)
a = input('What is the first number? ')
b = input('What is the second number? ')

# convert the strings to integers
a = int(a)
b = int(b)

# find the greatest common divisor
while a != b:
    print(a, b)  # see how and a and b change
    if a >= b:
        a = a - b
    else:
        b = b - a

# print the result
print("greatest common divisor:", a)

Please enter two integers.
I will calculate their GCD using Euclid's algorithm



ValueError: invalid literal for int() with base 10: '15.1'

Run the above program with these values:

- 10 and 9
- 100 and 99
- 7 and -5
- -7 and -5
- 3.14 and 2.77
- two and three

Don't worry if you don't understand all the details of the program yet. We will
get to them in the course.

In Python code, `#` marks the start of a **source code comment**, i.e. a note
meant for a human reading the source code. Good source code comments explain
tricky parts of the code.

## Finding Square Roots

Finding the square root of a number is a common task, and there are [a number of
different ways to calculate
it](https://en.wikipedia.org/wiki/Methods_of_computing_square_roots). One such
method is **Herons method**. To find the square root of a positive number $S$,
Heron's method creates a series of better-and-better approximations, $x_1, x_2,
x_3, \ldots$, until the approximation is good enough:

1. $x_1 = \frac{S}{2}$
2. $x_{i+1} = \frac{1}{2}(x_i + \frac{S}{x_i})$
3. Repeat step 2 until the guess is "good enough"

In the code below, "good enough" means that the difference in successive guesses
is less than 0.001. You can make the code run faster if you make this bigger,
e.g. 0.01 or 0.1, but then the answer will be less accurate.

In [2]:
S = input('Please enter a positive number: ')
S = float(S)  # convert the string to a float

x = S / 2             # initial guess
y = (x + S / x) / 2   # first approximation
# print(y)
while abs(y - x) > 0.001:
    x = y
    y = (x + S / x) / 2
    # print(y)

print('Square root of', S, 'is approximately', y)

Square root of 25.0 is approximately 5.000000000016778


## Programming Errors

Errors, or **bugs**, are common when writing computer programs. Even the best
and most experienced programmers make mistakes and need to watch for **bugs** in
their code.

In this course we will see three main kinds of programming errors: *syntax
errors*, *runtime errors*, and *logic errors*.

### Syntax Errors

A **syntax error** is a mistake in the spelling or grammar of your program.
Python catches some syntax errors *before* a program runs, but not all. Here's
an example of a syntax error:

  ```python
  print("oops!)   # syntax error: missing quote
  ```

### Runtime Errors

A runtime error is an error that occurs while your program is running. For
example, if you type the string `five` into Euclid's algorithm above, the
program will crash due to a runtime error.

### Logic Errors

A **logic error** is an error in the logic of your program.

For example, this version of the GCD algorithm has a logic error on line 4:

  1. start with two integers
  2. if they are the same, stop: you've found their GCD
  3. if they are different, replace the bigger number with the difference
     between it and the smaller number
  4. **go to step 3**  <-- **logic error**! Should go to step 2.

## Questions

1. What is an algorithm?

2. What is the GCD of 40 and 250? How many steps does Euclid's algorithm take to
   find it?

3. In the Python GCD program, there is this line:

   ```python
   if a > b:
   ```

   Will the program still work correctly if you change it to `if a >= b:`?

4. In the Python GCD program, there is this line:

   ```python
   while a != b:
   ```

   Will the program still work correctly if you delete the colon `:` at the end
   of the line?

5. For each of the following, say whether it is a *syntax error*, *runtime
   error*, or *logic error*:

   - `print "hello"`
   - in the GCD program, entering the numbers 15.1 and 39.2
   - in the GCD program, entering the numbers spelled as *five* and
     *thirty-nine*
   - in the square root program, replacing `while abs(y - x) > 0.001:` with `if
     abs(y - x) > 0.001`
   - in the square root program, entering the number 0
